### ISSUE TITLE
Trainer and QC defects found while preparing the first B5 run (record and remaining checks)

### ISSUE THREAD

**Issue by wyim-pgl** [wyim-pgl/transgenic#53]
**Goal**
Record the defects found on 2026-09-02/03 while preparing the first training run, so the fixes are auditable and the remaining checks are not lost.

Every one of these would have hit the first real training run. All are fixed and pushed; this issue is the record and the checklist of what still needs verifying on ACCESS.

**Fixed (commit)**
1. `--benchmark-steps` was accepted by the parser and passed to `train()` but never used: asking for a benchmark started a normal training run (81ee932).
2. The oversized-batch guard rejected every 129,024-nt tile: it compared `batch x seq_len` against a fixed 100,000 instead of the recipe's `max_encoder_seqlen` (81ee932).
3. `wandb` was imported at module load, so `--no-wandb` died with ModuleNotFoundError; neither the pgl-gpu environment nor the ACCESS container ships it (a20e4f4).
4. The startup banner read `gpu_props.total_mem`, which does not exist (`total_memory` does), crashing every run before the first batch, and `get_device_properties(0)` also threw on a CPU-only host (fea0d1d).
5. The validator compared v3 rows against the v2 token cap, reporting 410 false violations (47756f4).
6. `revision/scripts/62_geenuff_qc.py` produced zero flags against the real GeenuFF schema: error features carry no "error" in the name and reach transcripts through `association_transcript_piece_to_feature` (232ebe7).
7. Flag files key genes by GFF ID or Name while the builder generates its own keys for long or dotted ids, so GeenuFF and Swiss-Prot masking would have matched nothing for most species (22308ba).
8. The batch handler caught every exception and skipped the batch: the 129,024-nt tier dropped 1,093 of 1,103 batches while the loss curve looked healthy. Only CUDA OOM is caught now and the frozen-recipe path raises; the optimizer step and epoch loss count completed micro-batches instead of DataLoader indices (b5e4471, protocol A35).
9. The encoder load was reported in one stderr line and any failure continued with a randomly initialised encoder. The load is now audited per tensor, gated against a frozen expectation and written as `encoder_init_report.json` (b07e85f, protocol A34).

**Still to verify (mostly on ACCESS)**
- [ ] Confirm the realised global effective batch with Accelerate on more than one GPU equals the recipe's 96.
- [ ] Confirm the label cap: measure how many tiles exceed 8,192 tokens in the merged database and confirm nothing is silently truncated by the dataset's `maxlength`.
- [ ] Confirm the Hyena filter length actually initialises at `max_seq_len` 129,024 with a real forward and backward at that tier.
- [ ] Log loss, non-finite gradients and the global norm before and after clipping for the first few hundred optimizer steps.
- [ ] Check whether the final incomplete accumulation group is dropped without an optimizer step.
- [ ] Record container, CUDA, PyTorch and Transformers versions with the run.

Related: #18, #46, #49, #51.

**Comment by wyim-pgl** [comment]
Second checkbox done — "confirm the label cap: measure how many tiles exceed 8,192 tokens in the merged database and confirm nothing is silently truncated by the dataset's maxlength".

Measured against the frozen merged database (272,224 rows, `data/freeze/b5_full_v1.freeze.json`):

    largest stored label            5,920 tokens   (cap 8,192)
    rows at or over the cap             0
    label length histogram          <=2,048: 268,541 | 2,049-4,096: 3,575 | 4,097-8,000: 108

Stored counts come from `count_tokens_v3`, which is what the builder enforces the cap against — so I also ran the trainer's own `GFFTokenizer._tokenize` over every tile above 7,000 stored tokens (none exist) and a seeded random sample of 5,000 others. Longest tokenised label 4,198; nothing over the cap by either implementation. `isoformDataHyena.__getitem__` truncates over `self.maxlength` with no message (datasets.py:323-324, unlike the v2 path at :129-131 which at least warns), but it has nothing to truncate. Nothing is silently truncated.

The same comparison did surface a disagreement between the two implementations, on the `<empty>` label only — the tokenizer emits 2 tokens without `<s>` where the contract, the spec and the test all say 3 with it, across 25,159 tiles. Filed separately as #58 since it is a tokenizer question, not a cap question.

Raw output: `/home/pgl/scratch1/wyim/transgenic_data/b5/merged/label_cap_check.json`.

**Comment by wyim-pgl** [comment]
Fifth checkbox done — "check whether the final incomplete accumulation group is dropped without an optimizer step".

It is not dropped. `micro_done` (train_HyenaTransgenic.py:348) is initialised once before the epoch loop and never reset per epoch, so an incomplete group at an epoch boundary carries into the first optimizer step of the next epoch; evaluation runs under `no_grad` and leaves the gradients alone. Only the residual at the end of the final epoch is discarded — at most 95 micro-batches of 4.36 M.

Two things that follow from the same counter, neither of which changes the answer:

- `steps_per_epoch = ceil(len(train_dl) / accumulation_steps)` (:246) assumes each epoch starts a fresh group, so `t_total` over-counts. At 198,171 micro-batches, accumulation 96, 22 epochs: `t_total` is 45,430 against 45,414 actual optimizer steps, so the linear decay ends near 0.035 % of peak rather than 0.
- On resume `micro_done` restarts at 0 while `global_step` is restored, so accumulation-group boundaries shift at every resume. Correct behaviour — the partial group died with the process — but worth knowing when reading a loss curve across an A28 job chain.

Reading that loop turned up something that does matter, filed as #59: the training DataLoader is `shuffle=True` with no seedable sampler, so `accelerator.load_state` restoring the mid-epoch RNG gives the resumed epoch a **different permutation**, and the `step < resume_step` skip then drops the wrong samples. Measured on pgl-gpu (torch 2.5.1, accelerate 1.14.0): resuming at 20 of 64, 15 samples seen twice and 15 never seen; `use_seedable_sampler=True` gives 0 and 0. Mid-epoch resume is the normal path under A28 preemption, so that one wants fixing before the first seed.

**Comment by wyim-pgl** [comment]
Third checkbox, first half — "confirm the Hyena filter length actually initialises at `max_seq_len` 129,024".

It does, in all twelve layers. Built the model from `configs/b5_400m_win_v3.json` on CPU (414.3 M parameters) and inspected every sequence-length-sized tensor:

    layers 0..11: mixer.filter_fn.pos_emb.z  (1, 129024, 5)   seq_len attr 129024
    layers 0..11: mixer.filter_fn.pos_emb.t  (1, 129024, 1)   seq_len attr 129024
    transgenic.EncoderOutputPositionalEmbedding.pe  (129024, 768)

Twelve distinct `pos_emb` objects, no sharing. One trap for anyone repeating this: `model.named_buffers()` defaults to `remove_duplicate=True` and returned only a single layer, which reads as "eleven layers were left at the pretrained default". Pass `remove_duplicate=False`.

Forward passes on CPU, batch 1, no grad:

    30,720 nt  -> (1, 30720, 768)
    61,440 nt  -> (1, 61440, 768)
    129,024 nt -> (1, 129024, 768)

The A34 gate reported exactly the frozen expectation on the way through: `loaded 227/339 tensors (exact=80, adapted=147); missing=112; missing layers [8, 9, 10, 11]`.

The backward half of the checkbox is still open and still needs ACCESS: 129,024 nt does not fit the backward on a 24 GB card (measured, 1,093 of 1,103 batches OOM), so a real forward+backward at that tier can only be run on the A100/GH200. Script: `~/filter_len2.py` on pgl-gpu.
